src/scripts/list-groups.ts

```diff
@@ -1,82 +1,234 @@
+ * @param ma The `TaskEither` instance to handle.
+ * @returns {void}
+ * @throws {E} Throws the error contained in the `TaskEither` if it represents a
+ * failure.
+ */
+// eslint-disable-next-line functional/no-return-void
+const handleTaskEither: <E, A>(ma: TE.TaskEither<E, A>) => void = flow(
+  TE.tapIO(Console.info),
+  TE.getOrElse((error) => {
+    // eslint-disable-next-line functional/no-throw-statement, @typescript-eslint/no-throw-literal
+    throw error;
+  }),
+  (result) => void result()
+);
+
+/**
+ * Options passed to the `list-groups` command. Includes the client ID, private
+ * key, organisation URL, and user ID (optional).
+ *
+ * @example
+ * ```typescript
+ * const listGroupsOptions: ListGroupsOptions = {
+ *   clientId: '0oaddqhpa2nPVsxJX5d7',
+ *   privateKey: <private key>,
+ *   orgUrl: readonlyURL('https://dev-69870217.okta.com'),
+ *   userId: '00uddtlbyrKj9nyAX5d7',
+ * };
+ * ```
+ */
+type ListGroupsOptions = {
   readonly clientId: string;
   readonly privateKey: string;
   readonly orgUrl: ReadonlyURL;
-}> =>
+  readonly userId?: string;
+};
+
+/**
+ * Function which is used in the `list-groups` command.
+ */
+// eslint-disable-next-line functional/no-return-void
+const buildListGroupsFlow: (listGroupsOptions: ListGroupsOptions) => void =
+  flow(
+    parseListGroupsOptions,
+    createOktaClientAndUserIdOption,
+    createOktaGroupServiceAndUserIdOption,
+    getGroupsListTableString,
+    handleTaskEither
+  );
+
+/**
+ * Builds the `list-groups` command.
+ */
+// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
+export default (rootCommand: RootCommand): Argv<ListGroupsOptions> =>
   rootCommand.command(
     'list-groups',
     // eslint-disable-next-line quotes
-    "Provides a list of all groups' ID's, email addresses, display names, and statuses.",
-    // eslint-disable-next-line functional/no-return-void, functional/functional-parameters, @typescript-eslint/no-empty-function
-    () => {},
-    async (args: {
-      readonly clientId: string;
-      readonly privateKey: string;
-      readonly orgUrl: ReadonlyURL;
-    }) => {
-      const client = oktaReadOnlyClient(
-        {
-          clientId: args.clientId,
-          privateKey: args.privateKey,
-          orgUrl: args.orgUrl,
-        },
-        ['groups']
-      );
-      const service = new OktaGroupService(client);
-
-      const result = await groups(service)();
-
-      // eslint-disable-next-line functional/no-conditional-statement
-      if (E.isLeft(result)) {
-        // eslint-disable-next-line functional/no-throw-statement
-        throw result.left;
-      }
-    }
+    "Provides a list of all groups' IDs, email addresses, display names, and statuses. Allows a specification of a user ID to list only groups that the user is a member of.",
+    // eslint-disable-next-line functional/no-return-void, @typescript-eslint/prefer-readonly-parameter-types
+    (yargs) => {
```

Comment:
I think I would prefer this approach because it would make for a more intuitive CLI experience, where the user could potentially tab-complete from, say, the `user` command to see what user-related sub-commands are available, without having to read the main `--help` output.

Also, as more commands are added, the output of the main `--help` will become more overwhelming to the user. But if they can run `oktagon user --help`, the list of sub-commands shown will be fewer and therefore less overwhelming, and probably more relevant to what they want to do.

Another good reason to have `user` and `group` commands might be to be more consistent with the [Okta API scopes](https://developer.okta.com/docs/api/oauth2/). It feels slightly wrong to have the `list-groups --user` command in the `list-groups.ts` module, when it only requires the `okta.users.read` scope (`okta.groups.read` isn't required at all -- only to list all groups).

Then again, there are commands which have already been implemented which don't quite fit under `user` or `group`, such as `add-user-to-group` and `remove-user-from-group`. Although maybe these could appear as sub-commands of both `user` and `group`. I noticed they don't work without both the `okta.users.manage` **and** the `okta.groups.manage` scopes.

It would take a significant rewrite of the tool to conform to this design. Let me know if you would like me to create a ticket for the work.

Reply:
Yep, create a ticket for this.

Reply:
Have created this ticket: https://agiledigital.atlassian.net/browse/IE-361.